plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt

```diff
@@ -49,40 +60,78 @@ class FeatureDevSessionContext(val project: Project) {
+        val zippedProject = runBlocking { zipFiles(projectRoot) }
         val checkSum256: String = Base64.getEncoder().encodeToString(DigestUtils.sha256(FileInputStream(zippedProject)))
         return ZipCreationResult(zippedProject, checkSum256, zippedProject.length())
     }
 
-    fun ignoreFile(file: File): Boolean = try {
-        ignorePatternsWithGitIgnore.any { p -> p.containsMatchIn(file.path) }
-    } catch (e: Exception) {
-        true
+    private suspend fun ignoreFile(file: File): Boolean = coroutineScope {
+        val deferredResults = ignorePatternsWithGitIgnore.map { pattern ->
+            async {
+                pattern.containsMatchIn(file.path)
+            }
+        }
+        deferredResults.any { it.await() }
     }
 
-    fun ignoreFile(file: VirtualFile): Boolean = ignoreFile(File(file.path))
+    suspend fun ignoreFile(file: VirtualFile): Boolean = ignoreFile(File(file.path))
+
+    suspend fun zipFiles(projectRoot: VirtualFile): File = withContext(getCoroutineBgContext()) {
+        val files = mutableListOf<VirtualFile>()
+        VfsUtil.visitChildrenRecursively(
+            projectRoot,
+            object : VirtualFileVisitor<Void?>() {
+                override fun visitFile(file: VirtualFile): Boolean {
+                    if (file.isFile) {
+                        files.add(file)
+                        return true
+                    }
+                    return runBlocking {
+                        !ignoreFile(file)
+                    }
+                }
+            }
+        )
+
+        // Process files in parallel
+        val nonMatchingFiles = mutableListOf<VirtualFile>()
+        coroutineScope {
+            files.map { file ->
+                async {
+                    if (file.isFile && !ignoreFile(file)) {
+                        nonMatchingFiles.add(file)
+                    }
+                }
+            }.awaitAll()
+        }
 
-    private fun zipFiles(projectRoot: VirtualFile): File = createTemporaryZipFile {
-        VfsUtil.collectChildrenRecursively(projectRoot).map { virtualFile -> File(virtualFile.path) }.forEach { file ->
-            if (file.isFile() && !ignoreFile(file)) {
+        createTemporaryZipFileAsync { zipOutput ->
+            nonMatchingFiles.forEach { file ->
                 val relativePath = Path(file.path).relativeTo(projectRoot.toNioPath())
-                it.putNextEntry(relativePath.toString(), Path(file.path))
+                zipOutput.putNextEntry(relativePath.toString(), Path(file.path))
             }
         }
     }.toFile()
 
-    private fun parseGitIgnore(): List<String> {
+    private suspend fun createTemporaryZipFileAsync(block: suspend (ZipOutputStream) -> Unit): Path = withContext(EDT) {
+        val file = Files.createTempFile(null, ".zip")
+        ZipOutputStream(file.outputStream()).use { zipOutput -> block(zipOutput) }
+        file
+    }
+
+    private fun parseGitIgnore(): Set<String> {
```

Comment:
this logic makes the incorrect assumption that there is only a single `.gitignore` in the project. that is also a more involved change that should be done as a follow up

Reply:
sure